src/server/modules/audit/interfaces/LinkAuditService.ts

```diff
@@ -0,0 +1,61 @@
+import { UrlHistoryRecord } from './UrlHistoryRepository'
+
+export type LinkChangeKey =
+  | 'description'
+  | 'isFile'
+  | 'state'
+  | 'userEmail'
+  | 'longUrl'
```

Comment:
Something we could discuss about:

I was thinking it would be nice if we had a way to guarantee that every value here indeed corresponds to a key of `UrlType`. That way, in the future if someone accidentally renames or removes any key in `UrlType`, typescript compiler can crash and alert devs about the breaking change in this part of the codebase.

I started with this: `export type LinkChangeKey = keyof UrlType`. It gave the list of all possible keys in `UrlType`. We have now created an explicit dependency where anything of type `LinkChangeKey` must be a key of `UrlType`.

We now need to narrow the key space to just the 5 mentioned in your diff. The thing is, typescript syntax only allows for declaring supersets (using extends), but not subsets. Ideally we want to say that `UrlType extends LinkChangeKey { <additional keys here> }`, but this conflates link audit features with the pure data model types.

Instead, we declare our own 'reverse extend' type, like this: `type ReverseExtends = <U, T extends U> = T`. Whatever we feed into `U` will now be reverse-extended (i.e. a subset type)

As an example, `export type LinkChangeKey = ReverseExtends<keyof UrlType, 'description' | 'isFile'>`. Now, any variable with the `LinkChangeKey` type must be either 'description' or 'isFile', and we also have a guarantee that these strings are valid keys of `UrlType`. If you delete the `description` field of `UrlType`, tsc now crashes, saying `"TS2344: Type '"isFile" | "description"' does not satisfy the constraint 'keyof UrlType'.   Type '"description"' is not assignable to type 'keyof UrlType'."`

We're close, I just can't think of a good name that can convey this utility type for future engineers looking at your PR. Was thinking along the lines of `Certain` or `Specific`, so it reads naturally in english: `type LinkChangeKey = Specific<keyof UrlType, ...>`, reading literally as 'specific/certain keys of UrlType'.

Would like to seek your input on this matter as well, and any complications you foresee if we explicitly declare a dependency between `LinkChangeKey` and `UrlType`

Reply:
I think this is going in a great direction. I see no specific complications with declaring a dependency between `LinkChangeKey` and `UrlType`. I was also thinking about the same problem but I didn't tackle it as thoroughly as you until you wrote the above review.

Similar to you, initially I was stuck on how to select only certain key strings from `keyof UrlType`. However, thinking about it differently, we could use `Pick` to select the properties that are mutable from `UrlType` (`'description' | 'isFile' | 'state' | 'longUrl'`), before applying `keyof`.

Restating it more clearly, this would take the form 
```
type LinkChangeKey = keyof (Pick<UrlBaseType, 'description' | 'isFile' | 'state' | 'longUrl'>)
```

Using the native implementation has the benefit of being immediately recognisable for future developers, but I can also see benefits of explicitly defining a `ReverseExtends` utility type. 

What do you think?

Reply:
ooooh i like your idea, it doesn't rely on any strange magic and we don't need to introduce any additional utility types to achieve what we want. i vote to go with this approach 👍

Reply:
Updated in [c32e269](https://github.com/opengovsg/GoGovSG/pull/1910/commits/c32e26997d379dfafd67d6f3a410e67d35632858)

Reply:
lgtm! seeing the `userEmail` key being excluded feels bad, maybe LinkChangeKey can be derived from `UrlHistoryRecord` instead of `UrlBaseType` in the future 🤔  but looks good for now!